projects/RabbitMQ.Client/client/impl/ChannelBase.cs

```diff
@@ -992,7 +1038,46 @@ public ValueTask BasicPublishAsync<TProperties>(CachedString exchange, CachedStr
             }
 
             var cmd = new BasicPublishMemory(exchange.Bytes, routingKey.Bytes, mandatory, default);
-            return ModelSendAsync(in cmd, in basicProperties, body);
+            using Activity sendActivity = RabbitMQActivitySource.HasListeners ? RabbitMQActivitySource.Send(routingKey.Value, exchange.Value, body.Length) : default;
+
+            if (sendActivity != null)
+            {
+                BasicProperties props = PopulateActivityAndPropagateTraceId(basicProperties, sendActivity);
+                await ModelSendAsync(in cmd, in props, body);
+                return;
+            }
+
+            await ModelSendAsync(in cmd, in basicProperties, body);
+        }
+
+        private static BasicProperties PopulateActivityAndPropagateTraceId<TProperties>(TProperties basicProperties,
+            Activity sendActivity) where TProperties : IReadOnlyBasicProperties, IAmqpHeader
+        {
+            // This activity is marked as recorded, so let's propagate the trace and span ids.
+            if (sendActivity.IsAllDataRequested)
+            {
+                if (!string.IsNullOrEmpty(basicProperties.CorrelationId))
+                {
+                    sendActivity.SetTag(RabbitMQActivitySource.MessageConversationId, basicProperties.CorrelationId);
+                }
+            }
+
+            BasicProperties props = default;
+            if (basicProperties is BasicProperties properties)
+            {
+                props = properties;
+            }
+            else if (basicProperties is EmptyBasicProperty)
+            {
+                props = new BasicProperties();
+            }
+
+            var headers = props.Headers ?? new Dictionary<string, object>();
+
+            // Inject the ActivityContext into the message headers to propagate trace context to the receiving service.
+            DistributedContextPropagator.Current.Inject(sendActivity, headers, InjectTraceContextIntoBasicProperties);
```

Comment:
edge case: users might want to inject their own context (e.g. when a message is re-rerouted by the application code and context flows from somewhere else).

In this case we should link to the existing context and not inject new one.
Apparently we forgot to bring it from the OTEP, working on it - https://github.com/open-telemetry/semantic-conventions/pull/284/files#r1367340949

Reply:
If I'm understanding you correctly, I should only inject the context if one hasn't already been injected?

Reply:
> If I'm understanding you correctly, I should only inject the context if one hasn't already been injected?

Yes, correct. If any context is already present (added by the user, or by application code re-routing) it shouldn't be overwritten.

Reply:
yes, but also we should add a link with pre-existing context to the publish span. Clarified in the comment no OTel PR as well.

Reply:
Added this logic as well